[BugFix] AUTO_INCREMENT insert/load error when table has nullable column (#21846)

srlch opened the pull request:
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #21846

## Problem Summary(Required) ：
Problem:
Currently, we use nullExprInAutoIncrement to represent that if user specify null expression for auto increment column. But there is some bug for the nullExprInAutoIncrement

Insert:
If the table with auto increment column and other nullable column, the nullExprInAutoIncrement will be set even though the auto increment column is not set by null if user set null to the nullable column. This will cause insertion failure.

Load:
If the table with auto increment column and other nullable column. If user specify the NULL value for auto increment column explictly, and miss the nullable column, the load will be successfully finished but not cancel.This is unexpected behavior.

Solution:
1. Fix the nullExprInAutoIncrement problem.
2. refactor the code, throw exception when Insert/Update has nullExprInAutoIncrement == true

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 18 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/FileScanNode.java|0|2|00.00%|[363, 364]|
|:large_blue_circle:|com/starrocks/planner/StreamLoadScanNode.java|0|2|00.00%|[295, 296]|
|:large_blue_circle:|com/starrocks/sql/InsertPlanner.java|2|4|50.00%|[249, 267]|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|6|9|66.67%|[93, 94, 113]|
|:large_blue_circle:|com/starrocks/sql/UpdatePlanner.java|1|1|100.00%|[]|

srlch committed:
[BugFix] AUTO_INCREMENT insert/load error when table has nullable column

Problem:
Currently, we use nullExprInAutoIncrement to represent that if user specify
null expression for auto increment column. But there is some bug for the
nullExprInAutoIncrement

Insert:
If the table with auto increment column and other nullable column, the nullExprInAutoIncrement
will be set even though the auto increment column is not set by null. This will cause insertion failure.

Load:
If the table with auto increment column and other nullable column. If user specify the NULL value for
auto increment column explictly, and miss the nullable column, the load will be successfully finished
but not cancel.This is unexpected behavior.

Solution:
1. Fix the nullExprInAutoIncrement problem.
2. refactor the code, throw exception when Insert/Update has nullExprInAutoIncrement == true

Signed-off-by: srlch <linzichao@starrocks.com>

srlch commented on fe/fe-core/src/main/java/com/starrocks/sql/InsertPlanner.java:
```diff
@@ -263,8 +263,9 @@ private void castLiteralToTargetColumnsType(InsertStmt insertStatement) {
-                            nullExprInAutoIncrement = true;
+                        if (isAutoIncrement && row.get(idx).getType() == Type.NULL) {
+                            throw new SemanticException("AUTO_INCREMENT column: " + targetColumn.getName() +
+                                                        " can not be set to NULL");
```
done

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 18 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/FileScanNode.java|0|2|00.00%|[363, 364]|
|:large_blue_circle:|com/starrocks/planner/StreamLoadScanNode.java|0|2|00.00%|[295, 296]|
|:large_blue_circle:|com/starrocks/sql/InsertPlanner.java|2|4|50.00%|[249, 267]|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|6|9|66.67%|[113, 114, 136]|
|:large_blue_circle:|com/starrocks/sql/UpdatePlanner.java|1|1|100.00%|[]|
